test/buffer-test.c

```diff
@@ -272,6 +279,281 @@ buffer_initx_test(void)
     free(base);
 }
 
+static void
+buffer_initx_memalign_l2_test(void)
+{
+    enum buffer_set_opts opts = BUFSET_OPT_MEMALIGN_L2;
+    unsigned alignment = buffer_get_alignment(opts);
+    size_t buf_size = 268;
+    size_t buf_size_aligned = ALIGN_UP(buf_size, alignment);
+    size_t num_bufs = 100;
+    size_t region_size = ALIGN_UP(buf_size_aligned * num_bufs, alignment);
+    void *base = malloc(region_size);
+    NIOVA_ASSERT(base != NULL);
+
+    struct buffer_set bs = {0};
+
+    struct buffer_set_args bsa =
+    {
+        .bsa_set = &bs,
+        .bsa_opts = opts,
+        .bsa_nbufs = num_bufs,
+        .bsa_buf_size = buf_size,
+        .bsa_region = base,
+        .bsa_region_size = region_size,
+        .bsa_alignment = alignment,
+    };
+
+    uint8_t initialized = 1;
+    //either base is aligned on 1MB boundary or it is not
+    int rc = buffer_set_initx(&bsa);
+    NIOVA_ASSERT(rc == 0 ||
+                 (rc == -EINVAL && ((uintptr_t)base & (alignment - 1))));
+
+    // if base was not aligned, free it and allocate by using memalign
+    if (rc == -EINVAL)
+    {
+        free(base);
+        base = NULL;
+        initialized = 0;
+    }
+
+    if (!base)
+    {
+        rc = posix_memalign(&base, alignment, region_size);
+        NIOVA_ASSERT(rc == 0 && base);
+        bsa.bsa_region = base;
+    }
```

Comment:
I don't understand why we're providing a buffer here when `BUFSET_OPT_ALT_SOURCE_BUF` is not set.